unitycatalog/functions/client.py

```diff
@@ -0,0 +1,136 @@
+import json
+import logging
+import threading
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from dataclasses import dataclass
+from typing import Any, Dict, Literal, Optional
+
+_logger = logging.getLogger(__name__)
+
+_uc_function_client = None
+_client_lock = threading.Lock()
+
+
+@dataclass
+class FunctionExecutionResult:
+    """
+    Result of executing a function.
+    Value is always string, even if the function returns a scalar.
+    """
+
+    error: Optional[str] = None
+    format: Optional[Literal["SCALAR", "CSV"]] = None
+    value: Optional[str] = None
+    truncated: Optional[bool] = None
+
+    def to_json(self) -> str:
+        data = {k: v for (k, v) in self.__dict__.items() if v is not None}
+        return json.dumps(data)
+
+
+class BaseFunctionClient(ABC):
+    """
+    Base class for uc function calling client
+    """
+
+    def __init__(self, client: Any = None, **kwargs: Any) -> None:  # noqa: ARG002
+        self._lock = threading.Lock()
+
+    @abstractmethod
+    def create_function(self, function_info: Any) -> Any:
+        """Create a function"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def retrieve_function(self, function_name: str, **kwargs: Any) -> Any:
+        """Retrieve a function by its name"""
+        raise NotImplementedError
+
+    def validate_input_params(self, input_params: Any, parameters: Dict[str, Any]) -> None:
+        """
+        Validate passed parameters against the function's input parameters definition.
+
+        Args:
+            input_params (Any): The function's input parameters definition. It should be
+                of type InputParams in unitycatalog, or FunctionParameterInfos in databricks.
+            parameters (Dict[str, Any]): The parameters to validate.
+        """
+        parameters = deepcopy(parameters)
+        if input_params and input_params.parameters:
+            invalid_params: Dict[str, str] = {}
+            for param in input_params.parameters:
+                if param.name in parameters:
+                    try:
+                        self._validate_param_type(parameters.pop(param.name), param)
```

Comment:
do we need to use pop? can we use get?

Reply:
Ah it's because I check if there're leftovers in the parameters on line 72, and raise ValueError if there are.

Reply:
I see, I think I'd use the following:

```python
if keys := parameters.keys() - {p.name for p in input_params.parameters}:
    ...
```